spark3/src/main/java/org/apache/iceberg/spark/MigrateAction.java

```diff
@@ -0,0 +1,272 @@
+   * new location and further operations will not effect the source table.
+   *
+   * Use this if you would like to experiment with Iceberg without changing
+   * your original files.
+   * @param newLocation the base directory for the new Iceberg Table
+   * @return this for chaining
+   */
+  public MigrateAction withNewTableLocation(String newLocation) {
+    this.destDataLocation = newLocation;
+    this.destMetadataLocation = newLocation + "/" + ICEBERG_METADATA_FOLDER;
+    return this;
+  }
+
+  /**
+   * Adds additional properties to the newly created Iceberg Table. Any properties with
+   * the same key name will be overwritten.
+   * @param properties a map of properties to be included
+   * @return this for chaining
+   */
+  public MigrateAction withAdditionalProperties(Map<String, String> properties) {
+    this.additionalProperties.putAll(properties);
+    return this;
+  }
+
+  /**
+   * Adds an additional property to the newly created Iceberg Table. Any properties
+   * with the same key name will be overwritten.
+   * @param key the key of the property to add
+   * @param value the value of the property to add
+   */
+  public MigrateAction withAdditionalProperty(String key, String value) {
+    this.additionalProperties.put(key, value);
+    return this;
+  }
+
+  public MigrateAction(SparkSession spark, Identifier destTableName, Identifier sourceTableName) {
+    this.spark = spark;
+    this.sourceTableName = sourceTableName;
+    this.destTableName = destTableName;
+    this.destCatalog = getSparkSessionCatalogOrFail(spark, destTableName);
+
+    try {
+      sourcePartitionSpec = SparkSchemaUtil.specForTable(spark, Spark3Util.toTableIdentifier(sourceTableName));
+    } catch (AnalysisException e) {
+      throw new IllegalArgumentException("Cannot determining partitioning of " + sourceTableName.toString(), e);
+    }
+
+    try {
+      this.sourceTable =
+        spark.sessionState().catalog().getTableMetadata(Spark3Util.toTableIdentifier(sourceTableName));
+    } catch (NoSuchTableException | NoSuchDatabaseException e) {
+      throw new IllegalArgumentException(String.format("Could not find source table %s", sourceTableName), e);
+    }
+    validateSourceTable(sourceTable, ALLOWED_SOURCES);
+
+    String sourceTableProvider = sourceTable.provider().get().toLowerCase(Locale.ROOT);
+    this.sessionCatalogReplacement = !sourceTableProvider.equals("iceberg") && sourceTableName.equals(destTableName);
+
+    this.sourceTableLocation = CatalogUtils.URIToString(sourceTable.storage().locationUri().get());
+    this.destDataLocation = sourceTableLocation;
+    this.destMetadataLocation = sourceTableLocation + "/" + ICEBERG_METADATA_FOLDER;
+  }
+
+  private static void validateSourceTable(CatalogTable sourceTable, Set<String> supportedSourceTableProviders) {
+    String sourceTableProvider = sourceTable.provider().get().toLowerCase(Locale.ROOT);
```

Comment:
Does this mean that only Spark tables are supported and not Hive tables? I don't think that Hive tables have proviers.

Reply:
Added in a Hive test (at least I think so, used CREATE EXTERNAL TABLE LOCATION ....) which I though triggered the hive path, works for this.

Reply:
The provider for hive tables is "hive", doublechecked